src/lua/datetime.lua

```diff
@@ -915,6 +897,36 @@ local function datetime_parse_from(str, obj)
     end
 end
 
+local function datetime_cmp(lhs, rhs, is_raising)
+    if is_string(lhs) then
+        lhs = datetime_parse_full(lhs, '', 0, not is_raising)
+    end
+    if is_string(rhs) then
+        rhs = datetime_parse_full(rhs, '', 0, not is_raising)
+    end
+    if not is_datetime(lhs) or not is_datetime(rhs) then
+        if is_raising then
```

Comment:
Pasting the same comment here so it's easier to discuss:

> equality operators (== and ~=) are silently compare with strings
and if we could parse then we compare datetime values,
if we could not parse then silently return false;

But why? For example, decimal reports an error even in equality operators:
```
tarantool> dec.new(1) == 'abracadabra'
---
- error: '[string "return dec.new(1) == ''abracadabra''"]:1: incorrect value to convert
    to decimal as 2 argument'
...
```

Reply:
As later @sergepetrenko has discovered, we actually have inconsistency between `uuid` and `decimal`, one of which is not raising errors on invalid strings at comparison, but another is raising. 

```
tarantool> uuid.fromstr('744c5225-a962-40d2-a301-d401c43789eb') == 'x744c5225-a962-40d2-a301-d401c43789eb'
---
- false
...

tarantool> decimal.new(123) == 'x123'
---
- error: '[string "return decimal.new(123) == ''x123''"]:1: incorrect value to convert
    to decimal as 2 argument'
```

So this is up to @Mons to decide what is more idiomatic approach here?

Reply:
Have changed `==` and `~=` operators to be raising if there is incorrect datetime literal used on either side (it considered incorrect if it's not fully parsable)
```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -835,21 +835,24 @@ end
        date [T] time [ ] time_zone
     Returns constructed datetime object and length of accepted string.
 ]]
-local function datetime_parse_full(str, tzname, offset, is_quiet)
+local function datetime_parse_full(str, tzname, offset, parse_fully)
     check_str(str, 'datetime.parse()')
     local date = ffi.new(datetime_t)
     local len = builtin.tnt_datetime_parse_full(date, str, #str, tzname, offset)
     if len > 0 then
-        return date, tonumber(len)
-    elseif is_quiet then
-        return nil
+        if parse_fully and len ~= #str then
+            goto couldnot_parse_fully
+        else
+            return date, tonumber(len)
+        end
     elseif len == -builtin.TZ_NYI then
         error(("could not parse '%s' - nyi timezone"):format(str))
     elseif len == -builtin.TZ_AMBIGUOUS then
         error(("could not parse '%s' - ambiguous timezone"):format(str))
-    else -- len <= 0
-        error(("could not parse '%s'"):format(str))
     end
+    -- len <= 0 or partially parsed
+::couldnot_parse_fully::
+    error(("could not parse '%s'"):format(str))
 end
 
 --[[
@@ -899,10 +902,10 @@ end
 
 local function datetime_cmp(lhs, rhs, is_raising)
     if is_string(lhs) then
-        lhs = datetime_parse_full(lhs, '', 0, not is_raising)
+        lhs = datetime_parse_full(lhs, nil, 0, true)
     end
     if is_string(rhs) then
-        rhs = datetime_parse_full(rhs, '', 0, not is_raising)
+        rhs = datetime_parse_full(rhs, nil, 0, true)
     end
     if not is_datetime(lhs) or not is_datetime(rhs) then
         if is_raising then
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index 59ba07a6e..fd4ac8b75 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -194,11 +194,11 @@ test:test("Default date creation and comparison", function(test)
     test:is(ts1 > '1970-01-01', false, "not ts1 > 1970-01-1")
     test:is(ts1 >= '1970-01-01', true, "ts1 >= 1970-01-1")
 
-    -- equality comparisons are quiet for invalid strings
-    test:isnt(ts1 == 'x1970-01-01', "not ts1 == x1970-01-1")
-    test:isnt(ts1 ~= 'x1970-01-01', "not ts1 ~= x1970-01-1")
-
-    -- but other comparison operations are quite vocal
+    -- all comparisons operations are raising errors for invalid strings
+    assert_raises(test, couldnt_parse('x1970-01-01'),
+                  function() return ts1 == 'x1970-01-01' end)
+    assert_raises(test, couldnt_parse('x1970-01-01'),
+                  function() return ts1 ~= 'x1970-01-01' end)
     assert_raises(test, couldnt_parse('x1970-01-01'),
                   function() return ts1 < 'x1970-01-01' end)
     assert_raises(test, couldnt_parse('x1970-01-01'),
```